### PR TITLE
ERC-20 approval core functionality

### PR DESCRIPTION
**Pull request by ferencdg**
This PR only contains the core methods for the approval functionality. Interface methods, history will be added in later commits.

### CODE REVIEW

File: xtc/src/ledger.rs [hunk 1]
```diff
@@ -97,11 +181,38 @@ static InsufficientBalanceError: ErrorDetails = ErrorDetails {
...
+    msg: "Insufficient Allowance",
+    code: APIError::InsufficientAllowance,
+};
+
 static UnknownError: ErrorDetails = ErrorDetails {
     msg: "Unknown",
     code: APIError::Unknown,
 };
 
+// Disabled as the `name` clashes with a similar method in the cycles wallet
```
Comment (kanej) [inline xtc/src/ledger.rs:195]: Is this something we just have to do without?

File: xtc/src/ledger.rs [hunk 2]
```diff
@@ -284,37 +396,144 @@ mod tests {
         Principal::from_text("ai7t5-aibaq-aaaaa-aaaaa-c").unwrap()
     }
 
-    fn john() -> Principal {
+    fn charlie() -> Principal {
```
Comment (kanej) [inline xtc/src/ledger.rs:399]: Ha, I didn't realise I had briefly made it into the test suite.

File: xtc/src/meta.rs [hunk 3]
```diff
@@ -12,7 +12,7 @@ pub struct TokenMetaData<'a> {
 #[query]
 pub fn meta() -> TokenMetaData<'static> {
     TokenMetaData {
-        name: "Cycles",
+        name: "XTC Cycles",
```
Comment (kanej) [inline xtc/src/meta.rs:15]: We should keep this as cycles.

File: xtc/src/ledger.rs (hunk 2 above)
Comment (ferencdg) [reply xtc/src/ledger.rs:399]: :D fixed the problem

File: xtc/src/meta.rs (hunk 3 above)
Comment (ferencdg) [reply xtc/src/meta.rs:15]: changed back

File: .github/workflows/rust.yml [hunk 4]
```diff
@@ -50,7 +50,7 @@ jobs:
           key: cache-${{ runner.os }}-target-dir-cache-${{ hashFiles('cargo.lock') }}
 
       - name: Target dir clean
-        if: steps.target-dir-cache.outputs.cache-hit != 'true' || steps.target-clean-cache.outputs.cache-hit != 'true'
+        # if: steps.target-dir-cache.outputs.cache-hit != 'true' || steps.target-clean-cache.outputs.cache-hit != 'true'
```
Comment (qti3e) [inline .github/workflows/rust.yml:53]: The two previous steps are now not needed anymore if you don't want to cache `target`

File: xtc/src/ledger.rs (hunk 1 above)
Comment (qti3e) [reply xtc/src/ledger.rs:195]: I'd also say ERC-20's name should have higher priority. So let's uncomment this and comment out the cycles_wallet's name method.

File: xtc/src/ledger.rs [hunk 5]
```diff
@@ -97,11 +181,38 @@ static InsufficientBalanceError: ErrorDetails = ErrorDetails {
...
+
+#[query]
+fn decimal() -> u8 {
+    meta().decimal
+}
+
+#[query]
+fn totalSupply() -> Nat {
+    StatsData::get().supply
+}
```
Comment (qti3e) [inline xtc/src/ledger.rs:214]: I'd move this methods to `meta.rs` unless you want to change the directory structure as well.

File: xtc/src/ledger.rs (hunk 2 above)
Comment (qti3e) [reply xtc/src/ledger.rs:399]: @kanej You're also in `ic_kit` mock principals, so you're gonna be in a lot of people's test suites.

File: xtc/src/ledger.rs [hunk 6]
```diff
@@ -27,23 +31,102 @@ impl Ledger {
         self.balances.reserve(25_000 - self.balances.len());
     }
 
+    #[inline]
+    fn cleanupAllowances(&mut self, allower: &Principal, spender: &Principal) {
```
Comment (qti3e) [inline xtc/src/ledger.rs:35]: please use snake_case method names.

File: xtc/src/ledger.rs [hunk 7]
```diff
@@ -27,23 +31,102 @@ impl Ledger {
...
+    }
+
+    /***************************************************************************
+
+        1. Allower can allow more money to Spender than Allower's internal balance
+        2. Calling alowances twice replaces the previous allowance
+        3. Calling allownces with zero amount clears the allowance from the internal
+           map.
+
+    ***************************************************************************/
```
Comment (qti3e) [inline xtc/src/ledger.rs:65]: Please use cargo doc style comments, which is `///`

File: xtc/src/ledger.rs [hunk 8]
```diff
@@ -27,23 +31,102 @@ impl Ledger {
...
+
+        This method is allowed to violate the invariants of the ledger upon return:
+        for example by changing only the allowances and return an error without
+        changing the balance. The callers of this functions might use trap
+        to revert the state of the ledger.
+
+    ***************************************************************************/
+
+    #[inline]
+    pub fn transferFrom(
```
Comment (qti3e) [inline xtc/src/ledger.rs:87]: snake_case

File: xtc/src/ledger.rs [hunk 9]
```diff
@@ -12,6 +13,9 @@ use std::collections::HashMap;
 pub struct Ledger {
     // stores the cycle balance hold by the principal
     balances: HashMap<Principal, u64>,
+
+    // stores the allowances, approving Principal -> spender principal -> cycle balanace
+    allowances: HashMap<Principal, HashMap<Principal, u64>>,
```
Comment (qti3e) [inline xtc/src/ledger.rs:18]: Why not use a composit key `HashMap<(Principal, Principal), u64>`?

**Review by qti3e — CHANGES_REQUESTED** [review]
Overall it looks great, I didn't see any new update method exposed, but I think it should be part of another upcoming PR. I just made some suggestion and also some code formatting issues were pointed out.
Thank you!

File: .github/workflows/rust.yml (hunk 4 above)
Comment (ferencdg) [reply .github/workflows/rust.yml:53]: I don't want to completely remove them right now, I plan to revisit the CI as a speparate ticket as it has many problems like building in release mode unnecessarily.

File: xtc/src/ledger.rs (hunk 9 above)
Comment (ferencdg) [reply xtc/src/ledger.rs:18]: hmmm, I do like that idea, it would make the cleanup code simpler

Comment (ferencdg) [reply xtc/src/ledger.rs:18]: wait I think it is still better to keep the original datastructure as it is more space efficient:

Alice -> Bob1
Alice -> Bob2
Alice -> Bob3
Alice -> Bob4
....
Alice -> Bob10000

in this case we store Alice 10000 times using the datastructure you proposed

**Comment by ferencdg** [comment]
changes made, please merge

**Pull request merged**